File: VB/EditorStateExample.Module/CodeCentralProvider.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.DB.Helpers

Namespace WinWebSolution.Module
    'For demo purposes only!!!
    Public Class CodeCentralExampleInMemoryDataStoreProvider
        Public Const XpoProviderTypeString As String = "CodeCentralExampleInMemoryDataSet"
        Public Const ConnectionString As String = "XpoProvider=CodeCentralExampleInMemoryDataSet"
        Private Shared dataSet As DataSet
        Private Shared isRegistered As Boolean

        Shared Sub New()
            If (Not isRegistered) Then
                Try
                    dataSet = New DataSet()
                    DataStoreBase.RegisterDataStoreProvider(XpoProviderTypeString, New DataStoreCreationFromStringDelegate(AddressOf CreateProviderFromString))
                    isRegistered = True
                Catch
                    Throw New Exception(String.Format("Cannot register the {0}", GetType(CodeCentralExampleInMemoryDataStoreProvider).Name))
                End Try
            End If
        End Sub
        Public Shared Function CreateProviderFromString(ByVal connectionString As String, ByVal autoCreateOption As AutoCreateOption, <System.Runtime.InteropServices.Out()> ByRef objectsToDisposeOnDisconnect() As IDisposable) As IDataStore
            Dim result As New DataSetDataStore(dataSet, autoCreateOption)
            objectsToDisposeOnDisconnect = New IDisposable() { }
            Return result
        End Function
        Public Shared Sub Register()
        End Sub
    End Class
End Namespace

File: CS/EditorStateExample.Module/CodeCentralProvider.cs
using System;
using System.Data;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.DB.Helpers;

namespace WinWebSolution.Module {
    //For demo purposes only!!!
    public class CodeCentralExampleInMemoryDataStoreProvider {
        public const string XpoProviderTypeString = "CodeCentralExampleInMemoryDataSet";
        public const string ConnectionString = "XpoProvider=CodeCentralExampleInMemoryDataSet";
        private static DataSet dataSet;
        private static bool isRegistered;

        static CodeCentralExampleInMemoryDataStoreProvider() {
            if (!isRegistered) {
                try {
                    dataSet = new DataSet();
                    DataStoreBase.RegisterDataStoreProvider(XpoProviderTypeString, new DataStoreCreationFromStringDelegate(CreateProviderFromString));
                    isRegistered = true;
                }
                catch {
                    throw new Exception(string.Format("Cannot register the {0}", typeof(CodeCentralExampleInMemoryDataStoreProvider).Name));
                }
            }
        }
        public static IDataStore CreateProviderFromString(string connectionString, AutoCreateOption autoCreateOption, out IDisposable[] objectsToDisposeOnDisconnect) {
            InMemoryDataStore result = new InMemoryDataStore(dataSet, autoCreateOption);
            objectsToDisposeOnDisconnect = new IDisposable[] { };
            return result;
        }
        public static void Register() { }
    }
}